CRM/Contribute/BAO/Contribution.php

```diff
@@ -3172,6 +3172,9 @@ public static function recordFinancialAccounts(&$params, $financialTrxnValues =
 
     if (CRM_Utils_Array::value('contribution_mode', $params) == 'membership') {
       $isRelatedId = TRUE;
+      // CRM-21645: isRelatedId means we will check the id against the civicrm_contribution table
```

Comment:
@samuelsov instead of setting the ```$entityId``` here can we do it just above like at L3163 : 
```patch
...
 - elseif (!empty($params['membership_id'])) {
+ elseif (!empty($params['membership_id']) && CRM_Utils_Array::value('contribution_mode', $params) != 'membership') {
       //so far $params['membership_id'] should only be set coming in from membershipBAO::create so the situation where multiple memberships
       // are created off one contribution should be handled elsewhere
       $entityId = $params['membership_id'];
       $entityTable = 'civicrm_membership';
     }
     else {
+     $isRelatedId = (CRM_Utils_Array::value('contribution_mode', $params) == 'membership');
       $entityId = $params['contribution']->id;
       $entityTable = 'civicrm_contribution';
     }

- if (CRM_Utils_Array::value('contribution_mode', $params) == 'membership') {
-       $isRelatedId = TRUE;
- }
```

what do you think?